pytest_asyncio/plugin.py

```diff
@@ -484,7 +485,20 @@ def event_loop(request: "pytest.FixtureRequest") -> Iterator[asyncio.AbstractEve
     """Create an instance of the default event loop for each test case."""
     loop = asyncio.get_event_loop_policy().new_event_loop()
     yield loop
-    loop.close()
+    # Cleanup code copied from the implementation of asyncio.run()
```

Comment:
This new code works great, thank you! Users that don't define their own `event_loop` fixture will benefit immediately from the cleanup code.

Users that do override `event_loop` would have to add a lot of code to their fixture definition, though.

Would it be possible to perform the cleanup in `pytest_fixture_post_finalizer`, fixture? That way, users who override `event_loop` would only have to remove  the line `loop.close()` to benefit from the task and generator cleanup. It would be a much smoother transition and less error prone.

Reply:
Ok, I'll give it a go!

Reply:
I tried this and it seemed to work except that I got some warnings in two of the tox environments:

```
tests/test_simple.py ..Exception ignored in: <function BaseEventLoop.__del__ at 0x7f2fe9fc5550>
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 681, in __del__
    _warn(f"unclosed event loop {self!r}", ResourceWarning, source=self)
ResourceWarning: unclosed event loop <TestEventLoop running=False closed=False debug=False>
Exception ignored in: <socket.socket fd=18, family=AddressFamily.AF_UNIX, type=SocketKind.SOCK_STREAM, proto=0>
Traceback (most recent call last):
  File "/home/user/Documents/pytest-asyncio/.tox/pytest-min/lib/python3.9/site-packages/pluggy/callers.py", line 169, in _multicall
    for hook_impl in reversed(hook_impls):
ResourceWarning: unclosed <socket.socket fd=18, family=AddressFamily.AF_UNIX, type=SocketKind.SOCK_STREAM, proto=0>
Exception ignored in: <socket.socket fd=19, family=AddressFamily.AF_UNIX, type=SocketKind.SOCK_STREAM, proto=0>
Traceback (most recent call last):
  File "/home/user/Documents/pytest-asyncio/.tox/pytest-min/lib/python3.9/site-packages/pluggy/callers.py", line 169, in _multicall
    for hook_impl in reversed(hook_impls):
ResourceWarning: unclosed <socket.socket fd=19, family=AddressFamily.AF_UNIX, type=SocketKind.SOCK_STREAM, proto=0>
```

These are the pytest sessions for which I got this warning:
```
platform linux -- Python 3.9.9, pytest-6.1.0, py-1.11.0, pluggy-0.13.1
cachedir: .tox/pytest-min/.pytest_cache
rootdir: /home/user/Documents/pytest-asyncio, configfile: setup.cfg, testpaths: tests
plugins: asyncio-0.18.3.dev7+gdddf3e4.d20220326, flaky-3.7.0, hypothesis-6.39.4
asyncio: mode=auto
collected 77 items
```

```
platform linux -- Python 3.10.0, pytest-7.1.1, pluggy-1.0.0
cachedir: .tox/py310/.pytest_cache
rootdir: /home/user/Documents/pytest-asyncio, configfile: setup.cfg, testpaths: tests
plugins: asyncio-0.18.3.dev7+gdddf3e4.d20220326, flaky-3.7.0, hypothesis-6.39.4
asyncio: mode=auto
collected 77 items
```

Both times it happened in the same place, after `test_asyncio_marker`. I also noticed that it's using an event loop called `TestEventLoop` which is set up in `tests/respect_event_loop_policy/conftest.py`, so that is possibly leaking out of the tests that it's actually intended for. Not sure if that's a problem.

Reply:
> I tried this and it seemed to work except that I got some warnings in two of the tox environments

That sounds promising!

The `TestEventLoop` instance should not leak out of the `tests/respect_event_loop_policy` package.

It's hard for me to comment on the other two warnings, without getting my hands on the exact code you were running. Could you push the changes?

Reply:
The `ResourceWarning` was caused by `test_event_loop_scope.py::test_4` which sets an event loop to None without closing it. This seems to cause side-effects in other tests. I modified the test to close the loop before setting the current loop to None.

We should consider performing such tests with `pytester.runpytest_subprocess` to avoid side-effects on the rest of the test suite.